### PR TITLE
1.0dev

### PR DESCRIPTION
**Pull request by hanskrupakar**
Backend-agnostic model saver

### CODE REVIEW

File: fastestimator/trace/io/model_saver.py [hunk 1]
```diff
@@ -12,10 +12,84 @@
...
+        save_best (bool, str, optional): Best model saving monitor name. If True, the model loss is used. Defaults to
+            False.
+        save_best_mode (str, optional): Can be `'min'`, `'max'`, or `'auto'`. Defaults to 'max'.
+
+    """
+    def __init__(self, model: Union[tf.keras.Model, torch.nn.Module], model_name: str,
+                 save_dir: str, save_freq: int = 1, save_best: Union[bool, str]=False, save_best_mode:str='max'):
+
+        if save_best is True:
+            save_best = 'accuracy'
```
Comment (TortoiseHam) [inline fastestimator/trace/io/model_saver.py]: why would we assume 'accuracy' instead of something like loss? perhaps don't allow boolean as an argument for save_best

File: fastestimator/trace/io/model_saver.py [hunk 2]
```diff
@@ -12,10 +12,84 @@
...
+
+        if isinstance(model, tf.keras.Model):
+            self.ext = '.h5'
+        else:
+            self.ext = '.pt'
+
+        self.model = model
+        self.model_name=model_name
+        self.save_dir = save_dir
+        self.save_best = save_best
```
Comment (TortoiseHam) [inline fastestimator/trace/io/model_saver.py]: no need to store self.save_best since it will be in self.inputs already

File: fastestimator/trace/io/model_saver.py [hunk 3]
```diff
@@ -12,10 +12,84 @@
...
+        self.save_freq = save_freq
+        assert isinstance(self.save_freq, int), "save_freq must be integer"
+        if self.save_best_mode == "min":
+            self.best = np.Inf
+            self.monitor_op = np.less
+        elif self.save_best_mode == "max":
+            self.best = -np.Inf
+            self.monitor_op = np.greater
+        else:
+            raise ValueError("save_best_mode must be either 'min' or 'max'")
```
Comment (TortoiseHam) [inline fastestimator/trace/io/model_saver.py]: then why does it say we allow 'auto' in doc string?

File: fastestimator/trace/io/model_saver.py [hunk 4]
```diff
@@ -12,10 +12,84 @@
...
+        self.model = model
+
+    def on_begin(self, data: Data):
+        if self.save_dir:
```
Comment (TortoiseHam) [inline fastestimator/trace/io/model_saver.py]: this if check should be an assert in the init instead. no point in a model saver if there's no place to save the model

File: fastestimator/trace/io/model_saver.py [hunk 5]
```diff
@@ -12,10 +12,84 @@
...
+            self.save_dir = os.path.normpath(self.save_dir)
+            os.makedirs(self.save_dir, exist_ok=True)
+
+    def on_epoch_end(self, data: Data):
+        if isinstance(self.save_best, str):
```
Comment (TortoiseHam) [inline fastestimator/trace/io/model_saver.py]: this could just be 'if self.inputs' to remove the redundant member variable

File: fastestimator/trace/io/model_saver.py [hunk 6]
```diff
@@ -12,10 +12,84 @@
...
+        if isinstance(model, tf.keras.Model):
+            self.ext = '.h5'
+        else:
+            self.ext = '.pt'
+
+        self.model = model
+        self.model_name=model_name
+        self.save_dir = save_dir
+        self.save_best = save_best
+        self.save_best_mode = save_best_mode
```
Comment (TortoiseHam) [inline fastestimator/trace/io/model_saver.py]: don't need to store self.save_best_mode since you're generating self.monitor_op

File: fastestimator/trace/io/model_saver.py [hunk 7]
```diff
@@ -12,10 +12,84 @@
...
+        if self.save_dir:
+            self.save_dir = os.path.normpath(self.save_dir)
+            os.makedirs(self.save_dir, exist_ok=True)
+
+    def on_epoch_end(self, data: Data):
+        if isinstance(self.save_best, str):
+            if self.monitor_op(data[self.save_best], self.best):
+                self.best = data[self.save_best]
+                self._save_model("{}_best_{}{}".format(self.model_name, self.save_best, self.ext))
+                #self._save_model("{}_best_{}_{}.h5".format(self.model_name, self.save_best, str(self.best)))
```
Comment (TortoiseHam) [inline fastestimator/trace/io/model_saver.py]: remove dead code

File: fastestimator/trace/io/model_saver.py [hunk 8]
```diff
@@ -12,10 +12,84 @@
...
+        elif self.system.epoch_idx % self.save_freq == 0:
+            self._save_model("{}_epoch_{}_step_{}{}".format(
+                self.model_name, self.system.epoch_idx, self.system.batch_idx, self.ext))
+
+    def _save_model(self, name: str):
+        if self.save_dir:
```
Comment (TortoiseHam) [inline fastestimator/trace/io/model_saver.py]: no need for this if check - assert it in init

File: fastestimator/trace/io/model_saver.py [hunk 9]
```diff
@@ -12,10 +12,84 @@
...
 
 
 class ModelSaver(Trace):
-    def __init__(self, model):
-        pass
+    """Save trained model in hdf5/pt format.
+
+    Args:
+        model: Model instance defined by user (tf.keras.Model or torch.nn.Module)
+        model_name (str): Name of FE model.
```
Comment (TortoiseHam) [inline fastestimator/trace/io/model_saver.py]: remove type hints from function doc

**Comment by hanskrupakar** [comment]
@TortoiseHam you should really take a look at the model saver in the master branch. Most of the code patterns here are from there, I just spoofed the functionality....

Let me make some of the changes you talked about though.

**Comment by TortoiseHam** [comment]
@hanskrupakar we are trying to follow better code patterns now

**Comment by vbvg2008** [comment]
there are some apphub code mixed with this PR, closing this PR and I will make a new one about ModelSaver
